### PR TITLE
Check for "add to cart"

### LINKED ISSUE #203: Disable add to cart when size / color not selected
When the size / color pickers are not selected, the add to cart button must be disabled, as the size / color are mandatory to place order.

### PR DESCRIPTION
**Pull request by DavidCuentasMar**
fixes: #203 
yt link: https://www.youtube.com/watch?v=ewbDOXursUY

### CODE REVIEW

File: lib/screens/customer_product_detail_page.dart [hunk 1]
```diff
@@ -68,6 +61,38 @@ class _CustomerProductDetailPageState extends State<CustomerProductDetailPage> {
...
+    return bodyWidgets;
+  }
+
+  Widget _buildAddToCartFailedMessage() {
+    return Container(
+      alignment: AlignmentDirectional.center,
+      color: paletteRoseColor,
+      height: 75,
+      child: Text(
+        '色］［サイズ］を選択してください',
```
Comment (dragonfly91) [inline lib/screens/customer_product_detail_page.dart]: You can use this text:

when stockType = sizeAndColor, "色・サイズ　を選択してください"
when stockType = sizeOnly, "サイズを選択してください"
when stockType = colorOnly, "色を選択してください"

**Review by dragonfly91 — CHANGES_REQUESTED** [review]

File: lib/screens/customer_product_detail_page.dart (hunk 1 above)
Comment (DavidCuentasMar) [reply lib/screens/customer_product_detail_page.dart]: How the quantityOnly should behave? I found this case and not sure how to proceed

Comment (dragonfly91) [reply lib/screens/customer_product_detail_page.dart]: @DavidCuentasMar we don't need to show any text when stockType = quantityOnly 
These Japanese texts basically are alerting the user "please select size and / or color". This is not needed when only quantity is present. Does this answer your question?

File: lib/screens/customer_product_detail_page.dart [hunk 2]
```diff
@@ -312,17 +362,23 @@ class _CustomerProductDetailPageState extends State<CustomerProductDetailPage> {
...
-        .document(currentSignedInUser.email)
-        .get());
-
-    _updateCartModel(customer);
-
-    await db
-        .collection(constants.DBCollections.users)
-        .document(currentSignedInUser.email)
-        .updateData(customer.toMap());
+    if (_selectedItemColor != null && _selectedItemSize != null) {
```
Comment (dragonfly91) [inline lib/screens/customer_product_detail_page.dart]: I think this check is valid only for `stockType` = `sizeAndColor`. 

You can add a helper method like this:
```
bool _canAddToCart() {
  if (_product.stock.stockType == StockType.sizeAndColor) {
    return _selectedItemColor != null && _selectedItemSize != null;
  } else if (_product.stock.stockType == StockType.sizeOnly) {
    return _selectedItemSize != null;
  } else if (_product.stock.stockType == StockType.colorOnly) {
    return _selectedItemColor != null;
  }
  return true;
}
```
Then, you can use this method in the `GestureDetector` widget of `_buildAddToCartButton`:
```
if (_canAddToCart()) {
  _addToCart();
} else {
  setState(() {
    _showAddToCartFailedMessage = true;
  });
}
```
This way, you don't have to modify the `_addToCart` method.

File: lib/screens/customer_product_detail_page.dart [hunk 3]
```diff
@@ -68,6 +61,59 @@ class _CustomerProductDetailPageState extends State<CustomerProductDetailPage> {
...
+    );
+    bodyWidgets.add(SizedBox(height: 40));
+    bodyWidgets.add(_buildStockSelector());
+    bodyWidgets.add(SizedBox(height: 150));
+
+    return bodyWidgets;
+  }
+
+  Widget _buildAddToCartFailedMessage(StockType productStockType) {
+    String failMessage;
```
Comment (dragonfly91) [inline lib/screens/customer_product_detail_page.dart]: A better name could be `errorMessage`

File: lib/screens/customer_product_detail_page.dart [hunk 4]
```diff
@@ -266,10 +315,11 @@ class _CustomerProductDetailPageState extends State<CustomerProductDetailPage> {
       child: GestureDetector(
         onTap: () {
           _addToCart();
-
-          _scaffoldKey.currentState.showSnackBar(
-            _buildAddedToCartNotification(),
-          );
+          if (_selectedItemColor != null && _selectedItemSize != null) {
```
Comment (dragonfly91) [inline lib/screens/customer_product_detail_page.dart]: Please refer to my comment below. Snack bar should be shown whenever _addToCart succeeds:
```
if (_canAddToCart()) {
  _addToCart();
  _scaffoldKey.currentState.showSnackBar(      // <------- Add here!
    _buildAddedToCartNotification(),
  );
} else {
  setState(() {
    _showAddToCartFailedMessage = true;
  });
}
```

**Review by dragonfly91 — CHANGES_REQUESTED** [review]

**Comment by DavidCuentasMar** [comment]
In the last commit I removed one streambuilder (it was in _buildAppBar) and moved the other one to build method, this represents only making 1 api call and having access to product object in all widgets. Tell what you think about this.

**Review by dragonfly91 — APPROVED** [review]

**Pull request merged**